### DISCUSSION TITLE
PDF output with text

### DISCUSSION THREAD

**Discussion by OndrejSirocka** [chillerlan/php-qrcode#68] [Q&A] [answered]
We've been using the example class QRImageWithText to add a text below the QR code. Because we will need to print the QR codes, we will probably need to switch from PNG to PDF so that it's in vectors. Is it possible to add a text in PDF just like we do it in PNG? If so, would it be possible to add some sample class?

Thanks for your great work!

**Comment by codemasher** [comment]
Hey, i haven't done anything with the PDF output or FPDF itself yet and i'd recommend reading the [FPDF docs](http://www.fpdf.org/) in order to modify the PDF output (which is what i'd have to do by myself). If you set `QROptions::returnResource` to `true`, you'll get the FPDF object so that you can modify it.

**Comment by OndrejSirocka** [comment]
Thanks for the super fast response :-) I'll try to explore what the options are. I guess the basic idea is to extend QRFpdf class and add the text in the process of file creation.

Maybe outlining what should be in the dump function would be a great help because I'm getting a bit lost in all the dumping etc :-)

**Comment by codemasher** [comment, marked as answer]
Extending the output modules is quite easy. Assuming you want to modify the FPDF object *after* writing the QR Code, just do as follows:

```php
namespace My\App;

use chillerlan\QRCode\Output\QRFpdf;

class QRFpdfCustom extends QRFpdf{

	public function dump(string $file = null){
		// force set returnResource to true to get the FPDF object in the next step
		$this->options->returnResource = true;
		// retrieve the FPDF object
		$fpdf = parent::dump();

		// modify the PDF
#		$fpdf->Text($x, $y, 'text');

		// ...

		// proceed and mimic the output behaviour of the parent method
		$pdfData = $fpdf->Output('S');

		if($file !== null){
			$this->saveToFile($pdfData, $file);
		}

		if($this->options->imageBase64){
			$pdfData = sprintf('data:application/pdf;base64,%s', base64_encode($pdfData));
		}

		return $pdfData;
	}

}
```

However, if you need to modify the FPDF object on creation or while writing the QR Code, e.g. to modify size/position, you're better off copying the [whole `QRFpdf::dump()` method](https://github.com/chillerlan/php-qrcode/blob/be3beb936c21fe53a4e7e8f7f3582e9f02443666/src/Output/QRFpdf.php#L71-L110) to your custom class instead. To call the whole thing you'll need to specify `QRCode::OUTPUT_CUSTOM` and the FQCN of your custom class [like in the custom output example](https://github.com/chillerlan/php-qrcode/blob/be3beb936c21fe53a4e7e8f7f3582e9f02443666/examples/custom_output.php#L34-L35) and then call `QRCode::render()`.

**Comment by OndrejSirocka** [comment]
Makes sense, thanks a lot for your help!
